File: YOUPI-V2-backend/YOUPI/shared/events/src/main/kotlin/in/youpi/events/PushNotificationService.kt
package `in`.youpi.events

import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.Message
import com.google.firebase.messaging.Notification
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.math.BigDecimal

/**
 * Sends FCM push notifications directly to a user's device.
 *
 * Built specifically to close the "recharge confirmed but the app was
 * already closed" gap: emi_selection_screen.dart's synchronous polling
 * (recharge_viewmodel.dart, ~50s) and home_screen.dart's async follow-up
 * check (~75s more, only while Home is actually open) together cover most
 * cases, but if fulfillment takes even longer AND the user has since
 * closed the app entirely, neither of those can ever fire -- there's
 * nothing left running client-side to notice the eventual success. A push
 * notification is the only thing that can reach a closed app.
 *
 * The Firebase Admin SDK (FirebaseApp) is already initialized elsewhere
 * (see FirebaseConfig.kt) -- FirebaseMessaging.getInstance() just works,
 * no separate setup needed.
 */
@Service
class PushNotificationService {

    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * Fire-and-forget by design: a push failure (invalid/expired token,
     * FCM outage, whatever) must NEVER affect the recharge confirmation
     * itself, which has already fully succeeded by the time this is
     * called. Callers should not propagate exceptions from here -- this
     * function already catches everything and just logs.
     */
    suspend fun sendRechargeSuccessPush(
        fcmToken: String?,
        orderId: String,
        amountRupees: BigDecimal,
        earnedCoins: Int,
        earnedValueRupees: BigDecimal
    ) {
        if (fcmToken.isNullOrBlank()) {
            log.info("Push skipped for orderId={} -- no fcm_token on file for this user", orderId)
            return
        }

        try {
            val message = Message.builder()
                .setToken(fcmToken)
                .setNotification(
                    Notification.builder()
                        .setTitle("Recharge confirmed")
                        .setBody("Your ₹$amountRupees recharge went through — tap to see your reward")
                        .build()
                )
                // Data payload -- Flutter reads this on notification-tap
                // (and on a foreground message while the app is already
                // open) to play the SAME gold coin animation the
                // synchronous/async paths use, with the real numbers,
                // no extra API round-trip needed to figure out what was
                // earned.
                .putData("type", "recharge_success")
                .putData("orderId", orderId)
                .putData("amountRupees", amountRupees.toPlainString())
                .putData("earnedCoins", earnedCoins.toString())
                .putData("earnedValueRupees", earnedValueRupees.toPlainString())
                .build()

            // Firebase Admin SDK's send() is a blocking call -- must not
            // run on a coroutine dispatcher meant for non-blocking work.
            withContext(Dispatchers.IO) {
                FirebaseMessaging.getInstance().send(message)
            }
            log.info("Push sent for orderId={}", orderId)
        } catch (e: Exception) {
            // Deliberately swallowed -- see fun-level doc comment above.
            // A stale/invalid token (user reinstalled, uninstalled, etc.)
            // is the single most common failure here and is completely
            // expected; not worth surfacing as a real error.
            log.warn("Push send failed for orderId={} (non-fatal): {}", orderId, e.message)
        }
    }
}

File: YOUPI-V2-frontend/lib/core/services/push_notification_service.dart
// lib/core/services/push_notification_service.dart
//
// FCM push notification handling -- specifically built to close the last
// gap in the recharge-success reward animation flow:
//
//   1. emi_selection_screen.dart polls synchronously for ~50s right after
//      payment (recharge_viewmodel.dart's _pollOrderStatus).
//   2. If that times out, home_screen.dart's _checkPendingCoinAnimation
//      re-checks for ~75s more, but ONLY while Home is actually open.
//   3. If the recharge STILL hasn't confirmed by then, AND the user has
//      since closed the app entirely, nothing client-side is left running
//      to ever notice the eventual success -- there is no more "later" to
//      check on, because nothing is checking.
//
// A push notification is the only thing that can reach a fully closed
// app. The backend (RechargeService.kt) sends one the moment
// RECHARGE_SUCCESS is reached, carrying the exact same data the reward
// animation needs (orderId, amount, earnedCoins, earnedValue) in the
// payload -- no extra API round-trip required to figure out what to show.
//
// Usage: call PushNotificationService.init() once, early in main() (after
// Firebase.initializeApp(), before runApp()).

import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import '../../routes/app_router.dart';
import '../../data/repositories/user_repository.dart';
import 'coin_animation_signal.dart';
import 'storage_service.dart';

// Top-level function, NOT a class method -- the firebase_messaging plugin
// requires this exact shape (top-level or static, annotated
// @pragma('vm:entry-point')) because it can run in a separate isolate
// when a data message arrives while the app is fully terminated.
@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  // Deliberately does nothing beyond letting the OS show the notification
  // (which happens automatically for a message with a `notification`
  // block, even in this background isolate) -- the actual animation only
  // plays once the user taps it and the app comes to the foreground,
  // handled by onMessageOpenedApp/getInitialMessage below. Keeping this
  // handler minimal avoids needing Firebase re-initialized in a throwaway
  // background isolate just to show a coin animation nobody can see yet.
  debugPrint('FCM background message received: ${message.data}');
}

class PushNotificationService {
  static bool _initialized = false;

  static Future<void> init() async {
    if (_initialized) return;
    _initialized = true;

    FirebaseMessaging.onBackgroundMessage(firebaseMessagingBackgroundHandler);

    // iOS requires this explicitly; Android 13+ (API 33+) also requires a
    // runtime permission prompt now, same call handles both.
    final settings = await FirebaseMessaging.instance.requestPermission(
      alert: true,
      badge: true,
      sound: true,
    );
    debugPrint('FCM permission status: ${settings.authorizationStatus}');

    await _registerToken();
    FirebaseMessaging.instance.onTokenRefresh.listen((newToken) {
      UserRepository().updateFcmToken(newToken);
    });

    // App was fully terminated, user tapped the notification to open it.
    final initialMessage = await FirebaseMessaging.instance.getInitialMessage();
    if (initialMessage != null) {
      _handleRechargeSuccessMessage(initialMessage);
    }

    // App was backgrounded (not terminated), user tapped the notification.
    FirebaseMessaging.onMessageOpenedApp.listen(_handleRechargeSuccessMessage);

    // App was already open/foregrounded when the push arrived -- show the
    // animation immediately, no tap needed, since the user is right there.
    FirebaseMessaging.onMessage.listen(_handleRechargeSuccessMessage);
  }

  static Future<void> _registerToken() async {
    try {
      final token = await FirebaseMessaging.instance.getToken();
      if (token != null) {
        await UserRepository().updateFcmToken(token);
      }
    } catch (e) {
      // Non-fatal -- see UserRepository.updateFcmToken's own doc comment.
      debugPrint('FCM token registration failed (non-fatal): $e');
    }
  }

  static void _handleRechargeSuccessMessage(RemoteMessage message) {
    final data = message.data;
    if (data['type'] != 'recharge_success') return;

    final orderId = data['orderId'];
    final amountRupees = double.tryParse(data['amountRupees'] ?? '');
    if (orderId == null || amountRupees == null) {
      debugPrint('FCM recharge_success message missing/malformed data: $data');
      return;
    }

    // Reuses the EXACT SAME mechanism home_screen.dart's own async
    // follow-up check already uses (recharge_viewmodel.dart's
    // _stillProcessing path stores this same way) -- the push doesn't
    // trigger the animation directly itself, it just makes sure Home's
    // existing check has fresh data to find, then asks Home to look.
    // Since the backend only ever sends this push once the order has
    // ALREADY resolved to RECHARGE_SUCCESS, Home's very first status
    // check (no 15s wait needed) will resolve immediately.
    StorageService.setPendingCoinAnimation(orderId, amountRupees);

    AppRouter.router.go('/dashboard/home');
    CoinAnimationSignal.fire();
  }
}